src/basic-noise-qvm.lisp

```diff
@@ -0,0 +1,194 @@
+
+
+(defun (setf qubit-depolarization) (depol-prob qvm qubit)
+  "Save the kraus operators for the depolarizing channel on QUBIT. The depolarizing channel represents the average gate overrotation/underrotation noise. It is specified by DEPOL-PROB, which is the probability of overrotation/underrotation for the given QUBIT."
+  (let ((kraus-ops (generate-depolarizing-kraus-map depol-prob)))
+    (check-kraus-ops kraus-ops)
+    (setf (gethash qubit (depol-ops qvm)) kraus-ops)))
+
+
+(defun (setf qubit-fro) (fro qvm qubit)
+  "Save the readout povms for the fro provided on the specified qubit. The fR0 is assumed to be the average readout fidelity (avg of p(1|1) and p(0|0)) so as an approximation, we set p(0|0) = p(1|1) = fR0 in the readout povm assignment probabilities."
+  (setf (gethash qubit (readout-povms qvm)) (list fro (- 1.0d0 fro) (- 1.0d0 fro) fro)))
+
+
+(defmethod apply-all-kraus-maps ((qvm basic-noise-qvm) (instr quil:gate-application) kraus-ops)
+  ;; Apply all the kraus operators for each type of noise
+  ;; source. KRAUS-OPS is a list of ( (t1-operators) (t2-operators)),
+  ;; where t1-operators is a list of t1 noise maps for each qubit in
+  ;; the instruction.  If the instruction is a 2q gate, we need to
+  ;; tensor the kraus operators for each qubit's error source.
+  (loop :for noise-operators :in kraus-ops
+        :do (let ((simplified-kraus-ops  (reduce #'kraus-kron noise-operators)))
+              (when simplified-kraus-ops
+                (apply-kraus-map qvm instr simplified-kraus-ops)))))
+
+
+(defmethod transition ((qvm basic-noise-qvm) (instr quil:gate-application))
+  ;; For the current instruction in the program, apply the gate
+  ;; corresponding to that instruction, and then apply all the kraus
+  ;; operators for all the sources of noise following that
+  ;; instruction.
+  (let* ((gate   (pull-teeth-to-get-a-gate instr))
+         (params (mapcar (lambda (p) (force-parameter p qvm))
+                         (quil:application-parameters instr)))
+         (qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+         (t1-ops (loop :for q :in qubits :collect (gethash q (t1-ops qvm))))
+         (t2-ops (loop :for q :in qubits :collect (gethash q (t2-ops qvm))))
+         (depol-ops (loop :for q :in qubits :collect (gethash q (depol-ops qvm))))
+         (kraus-ops (list t1-ops t2-ops depol-ops)))
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple qubits) params)
+    (apply-all-kraus-maps qvm instr kraus-ops)
+    (incf (pc qvm))
+    qvm))
+
+
+(defmethod apply-classical-readout-noise ((qvm basic-noise-qvm) (instr quil:measure))
+  ;; Apply the classical readout noise to the state of an BASIC-NOISE-QVM
+  (%corrupt-qvm-memory-with-povm qvm instr (readout-povms qvm)))
+
+
+(defun tphi (t1 t2)
+  "Calculate t_phi from T1 and T2. t_phi = (2*t1*t2) / (2*t1 + t2). If both T1 and T2 are 0, TPHI = 0."
+  (if (and (plusp t1) (plusp t2))
```

Comment:
Use appropriately-typed constants: 2.0d0 instead of 2 and 0.0d0 instead of 0. Same thing everywhere else in the code.